### ISSUE TITLE
Task onError option not honored when workspace is present

### ISSUE THREAD

**Issue by wagoodman** [tektoncd/pipeline#4253]
# Expected Behavior
`onError` should be honored regardless of the presence of a workspace defined on a task.

# Actual Behavior
Task `onError` is not honored when a workspace is present on a task. Specifically, when specifying `onError: continue` for a task step, and the step has a non-zero return code,  then the task fails and non subsequent steps are executed.

# Steps to Reproduce the Problem

1. With `test.yaml` contents...:
  ```
  kind: Task
  apiVersion: tekton.dev/v1beta1
  metadata:
    name: test-taskrun-1
  spec:
  
    workspaces:
    - name: output
  
    steps:
      - image: alpine
        name: step0
        onError: continue
        script: |
          echo "This is step 0"
          ls -1R /tekton/steps/
          exit 1
  
      - image: alpine
        onError: continue
        script: |
          echo "This is step 1"
          ls -1R /tekton/steps/
          exit 2
      - image: alpine
        name: step2
        onError: continue
        script: |
          echo "This is step 2"
          ls -1R /tekton/steps/
          exit 3
  ---
  apiVersion: tekton.dev/v1beta1
  kind: Pipeline
  metadata:
    name: test-pipe
  spec:
  
    workspaces:
      - name: share
  
    tasks:
      - name: test-taskrun-1
        taskRef:
          name: test-taskrun-1
        workspaces:
          - name: output
            workspace: share
 ```
2. With the `pvc.yaml` contents...:
  ```
  apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    name: store
  spec:
    accessModes:
      - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
  ```
2. Load up the definitions: `kubectl apply -f test.yaml -f pvc.yaml`
3. Kick off the pipeline: `tkn pipeline start test-pipe -w name=share,volumeClaimTemplateFile=pvc.yaml`
4. Note that the task never gets past step0: `tkn pipelinerun logs test-pipe-<yourrunid> -f -n default`

If you take out the workspace in this example (which was pulled from the [original PR](https://github.com/pritidesai/pipeline/blob/fdeb0b1b38af6592e8aa5e2971959683706b08e7/docs/developers/README.md)) then the example works just fine.

# Additional Info

- Kubernetes version:
  ```
  Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"darwin/amd64"}
  Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T20:59:07Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version:
  ```
  Client version: 0.20.0
  Pipeline version: v20210920-0fcdd3962f
  Dashboard version: v20210920-b4aa2ff2ca
  ```

**Comment by pritidesai** [comment]
thanks @wagoodman for reporting the issue, I will look into it asap 🙏

**Comment by pritidesai** [comment]
Hey @wagoodman I have a fix for this issue in PR #4254

**Comment by wagoodman** [comment]
Thanks @pritidesai ! I went looking for the bug briefly with no result --nice work! I'll probably wait for this to hit nightly then try it out 👍
